### PR TITLE
[Enhancement]Added setting to limit parse large statement

### PR DESCRIPTION
**Pull request by packy92**
## What type of PR is this：
- [ ] bugfix
- [ ] feature
- [x] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

1. When parsing large sql, it costs a large amount of memory easily lead to FULL GC. We add a option `parse_tokens_limit` to prevent parsing sql exceeds the number of this option. I have tested its performance. This operation doesn't add extra time cost almostly.
2. In order to ensure compatibility, we use the old parser to parse sql again when the new parser throws an exception. When the old parser also failed, it swallowed the new parser exception contains meaningful error info but returned a useless error info provided by the old parser to client. This pr fixed this by throwing new parser exception when old parser encountered exception.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/TokenNumberListener.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+    public TokenNumberListener(int maxTokensNum) {
+        this.maxTokensNum = maxTokensNum;
+    }
+
+    @Override
+    public void visitTerminal(TerminalNode node) {
+        int index = node.getSymbol().getTokenIndex();
+        if (index >= maxTokensNum) {
+            throw new OperationNotAllowedException("Sql exceeds %d tokens, please consider modify parse_tokens_limit variable.",
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/parser/TokenNumberListener.java:19]: Normal users don't understand tokens concept, we should make the error message to be user-friendly.

**Comment by packy92** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 50 / 53 (94.34%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/parser/OperationNotAllowedException.java|2|4|50.00%|[10, 11]|
|:large_blue_circle:|com/starrocks/sql/parser/SqlParser.java|35|36|97.22%|[55]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/ConnectProcessor.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/parser/TokenNumberListener.java|8|8|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/TokenNumberListener.java (hunk 1 above)
Comment (packy92) [reply fe/fe-core/src/main/java/com/starrocks/sql/parser/TokenNumberListener.java]: have changed the error info more clear.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 49 / 52 (94.23%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/parser/OperationNotAllowedException.java|2|4|50.00%|[14, 15]|
|:large_blue_circle:|com/starrocks/sql/parser/SqlParser.java|35|36|97.22%|[55]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/ConnectProcessor.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/parser/TokenNumberListener.java|7|7|100.00%|[]|

**Comment by packy92** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 49 / 52 (94.23%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/parser/OperationNotAllowedException.java|2|4|50.00%|[14, 15]|
|:large_blue_circle:|com/starrocks/sql/parser/SqlParser.java|35|36|97.22%|[55]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/ConnectProcessor.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/parser/TokenNumberListener.java|7|7|100.00%|[]|

**Review by Seaven — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/SqlParser.java [hunk 2]
```diff
@@ -16,6 +17,55 @@
...
+        for (int idx = 0; idx < splitSql.size(); ++idx) {
+            String sql = splitSql.get(idx);
+            StatementBase statement = parseSingleSql(sql, sessionVariable);
+            statement.setOrigStmt(new OriginStatement(sql, idx));
+            statements.add(statement);
+        }
+        return statements;
+    }
+
+    public static StatementBase parseSingleSql(String sql, SessionVariable sessionVariable) {
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/parser/SqlParser.java:33]: what the different from other parse(line.69)?

Comment (packy92) [reply fe/fe-core/src/main/java/com/starrocks/sql/parser/SqlParser.java:33]: The old method only reveive a sqlMode parameter to define the property of  parser. I think there may be more paramter  in session variable for defining parser in the future. So I extract the define logic in a `setParserProperty()` method for expansibility.

**Comment by packy92** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 49 / 52 (94.23%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/parser/OperationNotAllowedException.java|2|4|50.00%|[14, 15]|
|:large_blue_circle:|com/starrocks/sql/parser/SqlParser.java|35|36|97.22%|[55]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/ConnectProcessor.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/parser/TokenNumberListener.java|7|7|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/SqlParser.java [hunk 3]
```diff
@@ -16,6 +17,55 @@
...
+
+    public static void setParserProperty(StarRocksParser parser, SessionVariable sessionVariable) {
+        StarRocksParser.sqlMode = sessionVariable.getSqlMode();
+        parser.removeErrorListeners();
+        parser.addErrorListener(new ErrorHandler());
+        parser.removeParseListeners();
+        parser.addParseListener(new TokenNumberListener(sessionVariable.getParseTokensLimit()));
+    }
+
     public static List<StatementBase> parse(String originSql, long sqlMode) {
```
Comment (HangyuanLiu) [inline fe/fe-core/src/main/java/com/starrocks/sql/parser/SqlParser.java:69]: please delete this

**Review by kangkaisen — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

File: fe/fe-core/src/main/java/com/starrocks/sql/parser/SqlParser.java (hunk 3 above)
Comment (packy92) [reply fe/fe-core/src/main/java/com/starrocks/sql/parser/SqlParser.java:70]: Replaced it in  most scenes. Some scenes is not easy to replace, so I have changed its implemention by invoke new parser method in its body.

**Review by HangyuanLiu — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Review by HangyuanLiu — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Comment by packy92** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 54 / 60 (90.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/scheduler/SqlTaskRunProcessor.java|0|1|00.00%|[28]|
|:large_blue_circle:|com/starrocks/sql/parser/OperationNotAllowedException.java|2|4|50.00%|[14, 15]|
|:large_blue_circle:|com/starrocks/analysis/CreateRoutineLoadStmt.java|8|10|80.00%|[615, 616]|
|:large_blue_circle:|com/starrocks/sql/parser/SqlParser.java|29|30|96.67%|[55]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMaterializedViewRefreshProcessor.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/http/rest/TableQueryPlanAction.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/common/SqlWithIdUtils.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/ConnectProcessor.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/parser/TokenNumberListener.java|7|7|100.00%|[]|

**Pull request merged**